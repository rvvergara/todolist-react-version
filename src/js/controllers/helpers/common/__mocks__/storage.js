const getDataFromLocalStorage = key => key;

export {
  getDataFromLocalStorage,
};
