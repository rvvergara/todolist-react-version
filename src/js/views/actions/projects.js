import uuid from 'uuid';

const getProjects = projects => ({
  type: 'GET_PROJECTS',
  projects,
});

const addProject = ({ name = 'PROJECT', description = 'SOME PROJECT' } = {}) => ({
  type: 'ADD_PROJECT',
  project: {
    id: uuid(),
    name,
    description,
    todos: [],
  },
});

const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id,
  updates,
});

const deleteProject = id => ({
  type: 'DELETE_PROJECT',
  id,
});

export {
  getProjects,
  addProject,
  updateProject,
  deleteProject,
};
