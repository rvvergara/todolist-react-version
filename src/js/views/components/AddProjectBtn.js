import React from 'react';
import PropTypes from 'prop-types';

const AddProjectBtn = (props) => {
  const {
    addProjectMode,
    clickAddProjectBtn,
  } = props;
  const displayClass = !addProjectMode ? 'btn btn-primary mt-1' : 'd-none';
  return (
    <button
      type="button"
      className={displayClass}
      onClick={clickAddProjectBtn}
    >
    Add New Project
    </button>
  );
};

AddProjectBtn.propTypes = {
  addProjectMode: PropTypes.bool.isRequired,
  clickAddProjectBtn: PropTypes.func.isRequired,
};

export default AddProjectBtn;
