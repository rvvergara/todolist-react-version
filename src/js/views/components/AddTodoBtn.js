import React from 'react';
import PropTypes from 'prop-types';

const AddTodoBtn = (props) => {
  const {
    handleTodoBtn,
  } = props;

  return (
    <button
      type="button"
      className="btn btn-primary btn-block"
      onClick={handleTodoBtn}
    >
    Add Todo
    </button>
  );
};

AddTodoBtn.propTypes = {
  handleTodoBtn: PropTypes.func.isRequired,
};

export default AddTodoBtn;
