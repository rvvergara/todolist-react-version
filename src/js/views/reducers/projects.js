export default (state = [], action) => {
  switch (action.type) {
    case 'GET_PROJECTS':
      return [...state, ...action.projects]

    case 'ADD_PROJECT':
    {
      const index = state.findIndex(project => project.name === action.project.name);
      return index === -1 ? state.concat(action.project) : state;
    }

    case 'UPDATE_PROJECT':
    {
      const { name } = action.updates;
      const conflictIndex = state.findIndex(project => project.name === name);
      const projectIndex = state.findIndex(project => project.id === action.id);
      const updatedProject = { ...state[projectIndex], ...action.updates };
      return conflictIndex === -1 ? [...state.slice(0, projectIndex), updatedProject, ...state.slice(projectIndex + 1)] : state;
    }
    case 'DELETE_PROJECT':
      return state.filter(project => project.id !== action.id);
    default:
      return state;
  }
};
