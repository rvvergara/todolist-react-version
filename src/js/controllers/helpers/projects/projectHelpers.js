import uuid from 'uuid';

import projectsController from '../../projectsController';

import * as localStorageData from '../common/storage';

export const createProjectsArray = () => {
  if (localStorageData.getDataFromLocalStorage('projectsArray') === null) {
    const projectsArray = [];
    localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
  }
};

// Create a default project for new users or after localstorage is cleared
export const createDefaultProject = () => {
  if (localStorageData.getDataFromLocalStorage('projectCount') === 0) {
    const defaultProject = projectsController.create('Default Project', uuid());
  }
  if (localStorageData.getDataFromLocalStorage('Default Project') !== null) {
    const defaultProject = localStorageData.getDataFromLocalStorage('Default Project');
  }
};

// Adding event listener to Project Form
// document.getElementById('projectNameForm').addEventListener('submit', e => {
//   e.preventDefault();
//   submitProjectForm(e.target);
// });
// // Adding event listener to Add Project Btn
// document.getElementById('addProjBtn').addEventListener('click', e => {
//   const action = 'new';
//   showProjectForm(e.target, action);
// });
