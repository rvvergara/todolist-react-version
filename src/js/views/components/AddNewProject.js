import React from 'react';
import { connect } from 'react-redux';
import AddProjectBtn from './AddProjectBtn';
import ProjectsForm from './ProjectsForm';
import {
  addProject,
} from '../actions/projects';
import {
  selectProject
} from '../actions/selectedProject';
import { addProjectModeSwitch, editProjectModeSwitch } from '../actions/projectForm';
import { addTodoModeSwitch, editTodoModeSwitch } from '../actions/todoForm';
import projectsController from '../../controllers/projectsController';

export class AddNewProject extends React.Component {
  state = {
    addProjectMode: false,
    error: '',
    projectName: '',
  };

  handleClickAddBtn = () => {
    this.props.addProjectModeSwitch();
    if(this.props.editProjectMode) this.props.editProjectModeSwitch();
    if(this.props.addTodoMode) this.props.addTodoModeSwitch();
    if(this.props.editTodoMode) this.props.editTodoModeSwitch();
  };

  handleChange = (key, val) => this.setState({
    [key]: val,
  });

  handleSubmitProjectForm = (e) => {
    e.preventDefault();
    const { addProject } = this.props;
    const name = this.state.projectName;
    const isValid = this.validateName(name);
    if(isValid && name){
      this.setState({
        projectName: '',
        error: '',
      });
      this.props.addProjectModeSwitch();
      const newProject = addProject({name});
      projectsController.create(name, newProject.project.id);
      this.props.selectProject(newProject.project);
      this.props.addNewProject(newProject.project);
    } else {
      this.setState({
        error: 'You have entered a duplicate or invalid name'
      });
    };
  };

  validateName = name => {
    const {projects} = this.props;
    return projects.findIndex(project => project.name === name) === -1; 
  };

  render() {
    return (
      <div>
        {!this.props.addProjectMode && <AddProjectBtn
          addProjectMode={this.state.addProjectMode}
          clickAddProjectBtn={this.handleClickAddBtn}
        />}
        { this.props.addProjectMode && <ProjectsForm
          addProjectMode={this.state.addProjectMode}
          submitProjectForm={this.handleSubmitProjectForm}
          handleChange={this.handleChange}
          error={this.state.error}
          name={this.state.projectName}
        />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  projects: state.projects,
  addProjectMode: state.projectForm.addProjectMode,
  editProjectMode: state.projectForm.editProjectMode,
  addTodoMode: state.todoForm.addTodoMode,
  editTodoMode: state.todoForm.editTodoMode,
});

export default connect(mapStateToProps, { addProject, addProjectModeSwitch, editProjectModeSwitch, selectProject, addTodoModeSwitch, editTodoModeSwitch})(AddNewProject);