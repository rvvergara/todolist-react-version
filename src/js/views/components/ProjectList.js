import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  getProjects,
} from '../actions/projects';
import { selectProject } from '../actions/selectedProject';
import ProjectItem from './ProjectItem';
import AddNewProject from './AddNewProject';
import {
  getDataFromLocalStorage,
} from '../../controllers/helpers/common/storage';

export class ProjectList extends React.Component {
  componentWillMount() {
    const {
      getProjects,
      selectProject,
    } = this.props;
    const projectsArray = getDataFromLocalStorage('projectsArray');
    if (projectsArray) {
      getProjects(projectsArray);
      selectProject(projectsArray[0].name);
    }
  }

  render() {
    const {
      projects,
    } = this.props;
    return (
      <div>
        <h2>Project List</h2>
        <div>
          {
          projects.map(project => (
            <ProjectItem
              project={project}
              key={project.id}
            />
          ))
        }
        </div>
        <AddNewProject />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  projects: state.projects,
  selectedProject: state.selectedProject,
});

ProjectList.propTypes = {
  projects: PropTypes.arrayOf(Object).isRequired,
  getProjects: PropTypes.func.isRequired,
  selectProject: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, { getProjects, selectProject })(ProjectList);
