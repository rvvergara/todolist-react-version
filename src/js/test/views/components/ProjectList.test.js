import React from 'react';
import { shallow } from 'enzyme';
import { ProjectList } from '../../../views/components/ProjectList';
import projects from '../../fixtures/projects';
import * as storage from '../../../controllers/helpers/common/storage';

describe('ProjectList', () => {
  let wrapper;
  let getProjects;
  let selectProject;

  beforeEach(() => {
    getProjects = jest.fn(() => ({ projects }));
    selectProject = jest.fn();
    storage.getDataFromLocalStorage = jest.fn(() => projects);
    wrapper = shallow(
      <ProjectList
        projects={projects}
        selectedProject={projects[0].name}
        getProjects={getProjects}
        selectProject={selectProject}
      />,
    );
  });

  describe('rendering and mounting', () => {
    test('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should call getDataFromLocalStorage', () => {
      expect(storage.getDataFromLocalStorage).toHaveBeenLastCalledWith('projectsArray');
    });

    test('should call getProjects action before mounting', () => {
      expect(getProjects).toHaveBeenLastCalledWith(projects);
    });

    test('should select first project before rendering', () => {
      expect(selectProject).toHaveBeenCalled();
    });

    test('should setState to projects', () => {
      expect(wrapper.state('projects')).toEqual(projects);
    });
  });

  describe('interacting with ProjectItem component', () => {
    test('it should handle delete', () => {
      const project = projects[0];
      wrapper.find('ConnectFunction').at(0).prop('handleDeleteProject')(project.id);
      expect(wrapper.state('projects')).toEqual([projects[1], projects[2]]);
      expect(selectProject).toHaveBeenLastCalledWith(projects[1]);
    });
  });

  describe('interacting with AddNewProject Component', () => {
    test('it should change state when adding new project', () => {
      const newProject = {
        name: 'Cool',
        id: 'aeha-2214',
      };
      wrapper.find('ConnectFunction').at(3).prop('addNewProject')(newProject);
      expect(wrapper.state('projects')).toEqual([...projects, newProject]);
    });
  });
});
