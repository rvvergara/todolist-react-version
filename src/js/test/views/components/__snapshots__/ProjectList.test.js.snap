// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList rendering and mounting should render correctly 1`] = `
<div>
  <h2>
    Project List
  </h2>
  <div>
    <ConnectFunction
      handleDeleteProject={[Function]}
      key="0"
      project={
        Object {
          "description": "Enim dolore quis tempor sit.",
          "id": 0,
          "name": "First",
          "todos": Array [
            Object {
              "id": 10,
            },
            Object {
              "id": 20,
            },
            Object {
              "id": 30,
            },
          ],
        }
      }
    />
    <ConnectFunction
      handleDeleteProject={[Function]}
      key="1"
      project={
        Object {
          "description": "Cupidatat occaecat anim excepteur dolore eu est ea aute ea commodo ipsum.",
          "id": 1,
          "name": "Second",
        }
      }
    />
    <ConnectFunction
      handleDeleteProject={[Function]}
      key="2"
      project={
        Object {
          "description": "Est qui nulla qui veniam nulla.",
          "id": 2,
          "name": "Third",
        }
      }
    />
  </div>
  <ConnectFunction
    addNewProject={[Function]}
  />
</div>
`;
